src/java.base/share/classes/java/util/WeakHashMap.java

```diff
@@ -240,6 +238,15 @@ public WeakHashMap() {
         this(DEFAULT_INITIAL_CAPACITY, DEFAULT_LOAD_FACTOR);
     }
 
+    /**
+     * use for calculate HashMap capacity, using default load factor 0.75
+     * @param size size
+     * @return HashMap capacity under default load factor
+     */
+    private static int calculateHashMapCapacity(int size){
+        return size + (size + 2) / 3;
```

Comment:
Consider integer overflow;  if size were Integer.MAX_VALUE / 2; the computation would overflow.
The negative result would eventually throw an exception. Using Long for the computation would avoid that and keep the expression simple.

Reply:
> if size were Integer.MAX_VALUE / 2; the computation would overflow

Actually will not, it must be slightly larger. it will only overflow when it be larger than Integer.MAX_VALUE * 0.75

But yes, it can overflow when there be a map as large as that.

> Using Long for the computation would avoid that and keep the expression simple.

but it be slower.

I do think a int check can do same thing, and would add it .

Thanks!

Reply:
@RogerRiggs

Hi. I added a overflow check.

The check makes sure it cannot overflow now.

I wrote a test to show this overflow check be correct.

```
public class A {

    /**
     * use for calculate HashMap capacity, using default load factor 0.75
     *
     * @param size size
     * @return HashMap capacity under default load factor
     */
    private static int calculateHashMapCapacity(int size) {
        if (size > (int) (Integer.MAX_VALUE * 0.75)) {
            return Integer.MAX_VALUE;
        }
        return size + (size + 2) / 3;
    }

    public static void main(String[] args) {
        for (int i = 1; i < Integer.MAX_VALUE ; ++i) {
            if (calculateHashMapCapacity(i) <= 0) {
                System.err.println(i);
                return;
            }
        }
    }
}

```

Reply:
Performance is a lesser issue. Given all of the other computation that occurs to populate the map, this computation is in the noise.  It is also likely that with more instructions to fetch and execute and a possible branch, the integer check is slower.
With current hardware, the long divide dominates the cost.  Addition is almost free.

Code maintainability is more important; keep the source code simple and concise.

Reply:
@RogerRiggs 
so you recommend `(int) Math.min(((m.size() * 4L + 2L) / 3L), Integer.MAX_VALUE)`? OK then, will change it.

Reply:
Works for me.  Thanks

Reply:
Just multiply by 0.75.

On a modern design, floating-point multiply is 4 clocks latency, 4 ops/clock throughput. FP max is 2 clocks latency, conversions int-float and vice versa 3 clocks latency, 4 ops/clock throughput.
Long division is 7-9 clocks, 2ops/clock throughput. Shift and add 2 clocks, 2/3 ops/clock througput. Compare is 1 clock, 3 ops/clock throughput, conditional move is 1 clock, 3 ops/clock throughput.

Seems like it's a wash.

Reply:
> Just multiply by 0.75.
> 
> On a modern design, floating-point multiply is 4 clocks latency, 4 ops/clock throughput. FP max is 2 clocks latency, conversions int-float and vice versa 3 clocks latency, 4 ops/clock throughput. Long division is 7-9 clocks, 2ops/clock throughput. Shift and add 2 clocks, 2/3 ops/clock througput. Compare is 1 clock, 3 ops/clock throughput, conditional move is 1 clock, 3 ops/clock throughput.
> 
> Seems like it's a wash.

no multiply but divide.

besides, did you add the cost for Math.ceil? it is the heaviest part.